### ISSUE TITLE
DL4J Keras Import: Add additional deconv test cases

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8317]
We don't have good enough Keras import coverage.

A few things to test:
* output padding configuration
* test cases where Keras adds implicit output padding (not sure why)

I don't believe we have any test cases with explicit (non-zero) output padding.

For example, Keras adds some implicit padding that does NOT appear in the JSON for the following deconv2d configurations:
k=2, inSize=3, stride=3, valid, d=1, nchw
k=1, inSize=8, stride=3, valid, d=1, nchw

i.e., Keras' output size does not match the (correct IMO) output size calculation here:
https://github.com/eclipse/deeplearning4j/blob/30b51f8085489c1519654039ada533c98c995837/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/util/ConvolutionUtils.java#L87-L88

For the former, Keras output size is 9, not 8; for the latter, 24 not 22. Even if we don't consider this strictly correct, we may consider mimicing this behaviour for import compatibility reasons.

**Comment by eraly** [comment]
I can do this! Can you assign me?

**Comment by AlexDBlack** [comment]
@eraly Done, thanks
You might as well start from this (which I used for op testing), but add non-zero output padding also
https://gist.github.com/AlexDBlack/d9d3c422a5c8c88aec10ab945964acc6

**Comment by eraly** [comment]
Looking at the Keras code, there is implicit padding for valid mode whenever the kernel size is less than the stride size, linked [here](https://github.com/keras-team/keras/blob/e8946d5240f3b18528d4d34668ee615907879953/keras/utils/conv_utils.py#L165)
So for these cases dl4j will give a dimension size that is smaller by (stride_size - kernel_size)

**Comment by eraly** [comment]
So it looks like this is the situation we need to think about -

output_padding is not specified in keras with valid mode and there are multiple possible deconv_sizes possible

Examples to illustrate:

```
from keras.layers import Conv2D, Conv2DTranspose, Input

i_in = [6,7,8,9,10,11,12]
#mode = ['same','valid']
mode = ['valid']
k = 2
s = 3

for m in mode:
    print("_____{}______".format(m))
    for i in i_in:
        conv = Conv2D(16, k, strides=s, padding=m)
        transpose_conv = Conv2DTranspose(1, k, strides=s, padding=m)
        input_a = Input(shape=(i, i, 1))
        x_conv_a = conv(input_a)
        x_transpose_a = transpose_conv(x_conv_a)

        print("Input shape: {}".format(keras.backend.int_shape(input_a)))
        print("Shape after convolution: {}".format(keras.backend.int_shape(x_conv_a)))
        print("Shape after transposed convolution: {}".format(keras.backend.int_shape(x_transpose_a)))
        print()

```

_____ valid ______
Input shape: (None, 6, 6, 1)
Shape after convolution: (None, 2, 2, 16)
Shape after transposed convolution: (None, 6, 6, 1)

Input shape: (None, 7, 7, 1)
Shape after convolution: (None, 2, 2, 16)
Shape after transposed convolution: (None, 6, 6, 1)

Input shape: (None, 8, 8, 1)
Shape after convolution: (None, 3, 3, 16)
Shape after transposed convolution: (None, 9, 9, 1)

Input shape: (None, 9, 9, 1)
Shape after convolution: (None, 3, 3, 16)
Shape after transposed convolution: (None, 9, 9, 1)

Input shape: (None, 10, 10, 1)
Shape after convolution: (None, 3, 3, 16)
Shape after transposed convolution: (None, 9, 9, 1)

Input shape: (None, 11, 11, 1)
Shape after convolution: (None, 4, 4, 16)
Shape after transposed convolution: (None, 12, 12, 1)

Input shape: (None, 12, 12, 1)
Shape after convolution: (None, 4, 4, 16)
Shape after transposed convolution: (None, 12, 12, 1)
